lib/recursive-methods.rb

```diff
@@ -1,49 +1,69 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
-
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return raise ArgumentError if n < 0
+  return 1 if n == 1 || n == 0
+  return n * factorial(n-1) if n > 1
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length == 1 || s.empty?
+  return s[-1] + reverse(s[1..-2]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) if I had gotten it to work
+# Space complexity: O(n)
 def reverse_inplace(s)
```

Comment:
1.  This method is not in place because `s[1..-2]` creates a new array.
2.  It's not working.

consider instead:

```ruby
def reverse_in_place(s, left = 0, right = s.length -1)
  if left < right
    temp = s[left]
    s[left] = s[right]
    s[right] = temp
    return reverse_in_place(s, left + 1, right - 1)
  end
  return s
end
```